### PR TITLE
GNN encoding/transformer

### PR DESCRIPTION
**Pull request by XUHAN314**
# Modulus Pull Request

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

File: constants.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    ckpt_name: str = "model.pt"
+
+    # training configs for sequence model
+    epochs_sequence: int = 200001
+    sequence_dim: int = 768
+    sequence_context_dim: int = 6
+    ckpt_sequence_path: str = "checkpoints/new_sequence"
+    ckpt_sequence_name: str = "sequence_model.pt"
+    sequence_batch_size: int = 1
+    produce_latents = True   #Set it as True when first produce latent representations from the encoder
```
Comment (hasethinvd) [inline constants.py:37]: change produce_latents = True to produce_latents: bool = True

File: test.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
+import torch
```
Comment (mnabian) [inline test.py:1]: Please move all the tests to a proper directory inside the parent `test` folder.

File: test_sequence.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
+import torch
```
Comment (mnabian) [inline test_sequence.py:1]: Same here

File: train.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
+import torch
```
Comment (mnabian) [inline train.py:1]: This should be moved to the `examples` directory.

File: constants.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
+import os
```
Comment (mnabian) [inline constants.py:1]: This should be moved to the `examples` directory.

File: dataset/meshPosition_all.txt
Comment (mnabian) [inline dataset/meshPosition_all.txt:1]: This should be moved to either the `examples` or the `modulus/datapipes` directory

File: modulus/datapipes/gnn/vortex_shedding_re300_1000_dataset.py [hunk 6]
```diff
@@ -0,0 +1,230 @@
+from dgl.data import DGLDataset
+from dgl.dataloading import GraphDataLoader
+import numpy as np
+import os
+import torch
+import dgl
+import pdb
```
Comment (mnabian) [inline modulus/datapipes/gnn/vortex_shedding_re300_1000_dataset.py:7]: Is this needed?

File: modulus/datapipes/gnn/vortex_shedding_re300_1000_dataset.py [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+from .utils import read_vtp_file, save_json, load_json
+from tqdm import tqdm
+
+class LatentDataset(DGLDataset):
```
Comment (mnabian) [inline modulus/datapipes/gnn/vortex_shedding_re300_1000_dataset.py:11]: Please add docstrings.

File: modulus/datapipes/gnn/vortex_shedding_re300_1000_dataset.py [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+		position_pivotal = None,
+		dist = None,
+		verbose=False
+	):
+		super().__init__(
+				name=name,
+				verbose=verbose,
+			)
+		self.split = split
+		self.sequence_len = 401
```
Comment (mnabian) [inline modulus/datapipes/gnn/vortex_shedding_re300_1000_dataset.py:29]: Can we make this configurable?

File: modulus/models/mesh_reduced/temporal_model.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+    def sample(self, z0, step_size, context=None):
+        z = z0  # .unsqueeze(1)
+
+        for i in range(step_size):
+            prediction = self.forward(z, context)[:, -1].unsqueeze(1)
+            z = torch.concat([z, prediction], dim=1)
+        return z
+
+    @staticmethod
+    def generate_square_subsequent_mask(
```
Comment (mnabian) [inline modulus/models/mesh_reduced/temporal_model.py:126]: please add docstrings

File: modulus/datapipes/gnn/vortex_shedding_re300_1000_dataset.py [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+    def __len__(self):
+        return len(self.z) // self.sequence_len
+
+    def __getitem__(self, idx):
+        return (
+            self.z[idx * self.sequence_len : (idx + 1) * self.sequence_len],
+            self.re[idx : (idx + 1)],
+        )
+
+    def get_re_number(self):
```
Comment (mnabian) [inline modulus/datapipes/gnn/vortex_shedding_re300_1000_dataset.py:47]: Please add docstrings

File: modulus/datapipes/gnn/vortex_shedding_re300_1000_dataset.py [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+                position_mesh,
+                position_pivotal,
+            )
+            z = z.reshape(1, -1)
+            record_z.append(z)
+        record_z = torch.cat(record_z, dim=0)
+        torch.save(record_z, "{}/latent_{}.pt".format(self.data_dir, self.split))
+
+
+class VortexSheddingRe300To1000Dataset(DGLDataset):
```
Comment (mnabian) [inline modulus/datapipes/gnn/vortex_shedding_re300_1000_dataset.py:115]: Please add docstrings

File: constants.py (hunk 1 above)
Comment (XUHAN314) [reply constants.py]: done.

File: train_sequence.py [hunk 12]
```diff
@@ -0,0 +1,278 @@
+import torch
```
[the remark this answers is not in the document]
Comment (XUHAN314) [reply train_sequence.py]: done.

File: constants.py (hunk 5 above)
Comment (XUHAN314) [reply constants.py]: done.

File: modulus/datapipes/gnn/vortex_shedding_re300_1000_dataset.py (hunk 6 above)
Comment (XUHAN314) [reply modulus/datapipes/gnn/vortex_shedding_re300_1000_dataset.py:18]: delete pdb.

File: modulus/datapipes/gnn/vortex_shedding_re300_1000_dataset.py [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+		nuAll = 1/ReAll
+		listCatALL = []
+		for i in range(3):
+			re = ReAll**(i+1)
+			nu = nuAll**(i+1)
+			listCatALL.append(re/re.max())
+			listCatALL.append(nu/nu.max())
+		if self.split == "train":
+			index = [i for i in range(101) if i % 2 == 0]
+			#index = [0]
```
[the remark this answers is not in the document]
Comment (XUHAN314) [reply modulus/datapipes/gnn/vortex_shedding_re300_1000_dataset.py]: done.

File: modulus/datapipes/gnn/vortex_shedding_re300_1000_dataset.py [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+		else:
+			index = [i for i in range(101) if i % 2 == 1]
+		#index_interest = [i for i in range(101) if i % 2 == 0]
```
[the remark this answers is not in the document]
Comment (XUHAN314) [reply modulus/datapipes/gnn/vortex_shedding_re300_1000_dataset.py]: done.

File: modulus/datapipes/gnn/vortex_shedding_re300_1000_dataset.py [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+		
+		self.split = split
+		self.rawData = np.load(os.path.join(self.data_dir,'rawData.npy'),
+							   allow_pickle = True)
+		
+
+        #select training and testing set
+		if self.split == "train":
+			self.sequence_ids = [i for i in range(101) if i%2 == 0]
+			#self.sequence_ids = [0]
```
[the remark this answers is not in the document]
Comment (XUHAN314) [reply modulus/datapipes/gnn/vortex_shedding_re300_1000_dataset.py]: done.

File: modulus/datapipes/gnn/vortex_shedding_re300_1000_dataset.py [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+		if self.split == "test":
+			self.sequence_ids = [i for i in range(101) if i%2 == 1]
+
+
+		# solution states are velocity and pressure
+		self.solution_states = torch.from_numpy(self.rawData['x'][self.sequence_ids,:,:,:]).float()
+
+		# cell volume
+		#self.M = torch.from_numpy(self.rawData['mass'])
```
[the remark this answers is not in the document]
Comment (XUHAN314) [reply modulus/datapipes/gnn/vortex_shedding_re300_1000_dataset.py]: done.

File: modulus/datapipes/gnn/vortex_shedding_re300_1000_dataset.py [hunk 17]
```diff
@@ -0,0 +1,230 @@
...
+		 						self.edge_stats["edge_std"])
+	
+	def __len__(self):
+		return self.sequence_len * self.sequence_num
+	
+	def __getitem__(self, idx):
+		sidx = idx // self.sequence_len
+		tidx = idx % self.sequence_len
+		
+		# node_features = torch.cat([self.solution_states[sidx,tidx],
```
[the remark this answers is not in the document]
Comment (XUHAN314) [reply modulus/datapipes/gnn/vortex_shedding_re300_1000_dataset.py]: done.

File: modulus/datapipes/gnn/vortex_shedding_re300_1000_dataset.py [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+		denormalized_invar = invar * std + mu
+		return denormalized_invar
+	
+if __name__ == '__main__':
+	data_set = VortexSheddingRe300To1000Dataset(split = 'train')
+	data_loader = GraphDataLoader(data_set,
+								  batch_size=5,
+								  shuffle=True)
+	for batched_graph in data_loader:
+		pdb.set_trace()
```
[the remark this answers is not in the document]
Comment (XUHAN314) [reply modulus/datapipes/gnn/vortex_shedding_re300_1000_dataset.py]: Yes. Removed.

File: modulus/datapipes/gnn/vortex_shedding_re300_1000_dataset.py (hunk 7 above)
Comment (XUHAN314) [reply modulus/datapipes/gnn/vortex_shedding_re300_1000_dataset.py:27]: Done.

**Comment by mnabian** [comment]
@XUHAN314 please black format your code.

**Comment by mnabian** [comment]
@hasethinvd could you please confirm that you have tested the latest version of the code and everything works fine?

**Comment by hasethinvd** [comment]
Yes,  it works fine Mohammad

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Review by mnabian — APPROVED** [review]

**Review by mnabian — APPROVED** [review]

**Comment by ktangsali** [comment]
Can we please add description to this PR?

**Pull request merged**